tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,158 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow.op;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * A class to manage scoped (hierarchical) names for operators.
+ *
+ * <p>{@code NameScope} manages hierarchical names where each component in the hierarchy is
+ * separated by a forward slash {@code '/'}. For instance, {@code nn/Const_72} or {@code
+ * nn/gradient/assign/init}. Each scope is a subtree in this hierarchy.
+ *
+ * <p>Use {@code NameScope} to group related operations within a hierarchy, which for example lets
+ * tensorboard coalesce nodes for better graph visualizations.
+ *
+ * <p>This class is package private, user code creates {@link Scope} which internally delegates
+ * calls to an underlying {@code NameScope}.
+ *
+ * <p>This class is thread-safe.
+ */
+final class NameScope {
+
+  NameScope withSubScope(String scopeName) {
+    if (baseName == null) {
+      checkPattern(OP_NAME_REGEX, scopeName);
+    } else {
+      checkPattern(SUBSCOPE_NAME_REGEX, scopeName);
+    }
+
+    if (baseOpName != null) {
+      // Use the base name instead to derive the subscope.
+      scopeName = baseOpName;
```

Comment:
This is a bit confusing in that `withSubScope` ignores the argument if `baseOpName` is `null`.
One alternative would be that `withSubScope` ignores the `baseOpName` and just creates a new sub-scope.
So something like:

```java
Const.create(s.withName("foo"), ...) // Results in "foo"
Const.create(s.withName("foo").withSubScope("foo"), ...)  // Results in "foo/Const" - as if the withName was ignored
```

That way it is clearer than `withName` only applies to the name (suffix) of the op and doesn't affect the scoping prefixes ever.

What do you think?